### PR TITLE
Added listener for Fused Location Provider.

### PR DESCRIPTION
**Pull request by realbigmitchell**
I added a listener for the Fused Location Provider for Android. This allows users to easily collect higher-fidelity reference data for algorithm development, since the default Android GPS location provider does not fuse information from the IMU or alti-barometer onboard the device. The implementation is backwards-compatible with any applications reading the log data, such as the MATLAB GNSS Analysis app, because the Fused Location messages follow the same format as the existing GPS location update messages. I tested this by collecting data and plotting it with the MATLAB application.

### CODE REVIEW

**Comment by realbigmitchell** [comment]
@googlebot I signed it!

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 import android.os.Bundle;
 
 /**  A class representing an interface for logging GPS information. */
-public interface GnssListener {
+public interface SensorFusionListener {
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java:28]: It sounds like this is a listener for any kind of measurement, whether it be GNSS, Sensor, Fused, or so on- consider renaming it along the lines of `MeasurementListener` and updating the class javadoc to reflect that, i.e. "A class representing an interface for logging a measurement."

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java [hunk 2]
```diff
@@ -251,6 +281,24 @@ public void registerLocation() {
...
+  public void unregisterLocation() {
+    mLocationManager.removeUpdates(mLocationListener);
+  }
+
+  public void registerFusedLocation() {
+    LocationRequest locationRequest = new LocationRequest();
+    locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
+    locationRequest.setInterval(1000);
+    locationRequest.setFastestInterval(100);
+    LocationServices.FusedLocationApi.requestLocationUpdates(mGoogleApiClient, locationRequest, mFusedLocationListener);
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java:293]: In this class you've already got mLocationProvider as a field, and in another PR you're adding mSensorManager as a field. For symmetry, make the the FusedLocationClient a field as well. Instantiate it in the class constructor, like `mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)`.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java [hunk 3]
```diff
@@ -21,19 +21,21 @@
 import android.location.GnssNavigationMessage;
 import android.location.GnssStatus;
 import android.location.Location;
-import android.location.LocationListener;
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java:24]: Why did you remove this import? If you restore it, you can remove the `android.location.` qualifier on LocationManager in the rest of the code.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java [hunk 4]
```diff
@@ -21,19 +21,21 @@
...
 import java.util.concurrent.TimeUnit;
+import com.google.android.gms.common.api.GoogleApiClient;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationServices;
 
 /**
  * A container for GPS related API calls, it binds the {@link LocationManager} with {@link UiLogger}
  */
-public class GnssContainer {
+public class SensorFusionContainer {
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java:38]: I think this class can have a more appropriate name. What this class appears to do is provide measurements from a few measurement providers - GNSS, sensor, fused - to an array of listeners. Why not call it something along the lines of `MeasurementProvider`? Also please update the javadoc to reflect that this class binds with many providers, not just GNSS anymore, and that the listeners are any instances of `MeasurementListener`, not just `UiLogger`.

**Review by alexanderkuznetsov96 — CHANGES_REQUESTED** [review]

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java (hunk 1 above)
Comment (alexanderkuznetsov96) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java:28]: Also, since we'll now be getting locations from both GNSS and FLP, you'll want to add a second argument to `onLocationChanged` that specifies the source of the location as a string to discern between locations from the two providers, like `onLocationChanged(Location location, String source)`. Then in your callback, you can log the location from FLP with `logger.onLocationChanged(location, "FLP");`, for example.

**Review by alexanderkuznetsov96 — CHANGES_REQUESTED** [review]

**Comment by alexanderkuznetsov96** [comment]
Mitchell, thanks for contributing to this project. The follow-up PR adding IMU seems to have a lot of overlap with this PR, so let's merge this one before I review the next.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java (hunk 3 above)
Comment (realbigmitchell) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java:24]: I removed this import because both android.location and com.google.android.gms.location have differing implementations of the LocationListener class used in this file. I figured using the full name for both classes throughout the file would keep things consistent since only one can be imported and make it easier to distinguish which LocationListener implementation was being referenced at any given time. Would you like me to put the import back for the android.location version and remove the qualifier throughout the rest of the file? Happy to do so if you think that makes more sense.

Comment (alexanderkuznetsov96) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java:24]: Oh, no that makes sense. Thanks!

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java (hunk 4 above)
Comment (realbigmitchell) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java]: I struggled with what to call this class as well. Thanks for the suggestion, it's been implemented!

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java (hunk 1 above)
Comment (realbigmitchell) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java]: > It sounds like this is a listener for any kind of measurement, whether it be GNSS, Sensor, Fused, or so on- consider renaming it along the lines of `MeasurementListener` and updating the class javadoc to reflect that, i.e. "A class representing an interface for logging a measurement."

Done!

Comment (realbigmitchell) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java]: > Also, since we'll now be getting locations from both GNSS and FLP, you'll want to add a second argument to `onLocationChanged` that specifies the source of the location as a string to discern between locations from the two providers, like `onLocationChanged(Location location, String source)`. Then in your callback, you can log the location from FLP with `logger.onLocationChanged(location, "FLP");`, for example.

The Android Location class passed in the first argument already has a getProvider() method which is used by the MeasurementListener implementations to access and log this information. Would you still like me to add it as a separate argument? Happy to do so if you'd like!

https://developer.android.com/reference/android/location/Location#getProvider()

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java (hunk 2 above)
Comment (realbigmitchell) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java]: I agree that this suggestion makes the class implementation more consistent. Unfortunately it looks like the FusedLocationProviderClient has bugs before Google Play services version 11.6.0:

https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient

I tried changing the Gradle configuration to upgrade play-services-location to version 11.6.0, but I wasn't able to build successfully until I upgraded both play-services-location and play-services-maps to version 12.0.0. 

It looks like both the FusedLocationProviderApi and ActivityRecognitionApi were deprecated starting in version 12, so to implement your suggestion I believe I'd have to convert both to their respective Client versions. FusedLocationProviderClient has different method signatures for requestLocationUpdates() than FusedLocationProviderApi, so if I'm not missing something (which I definitely could be), this will significantly increase the scope of this change/PR.

Would you like me to go ahead and make those changes? If so, should I add more commits to this PR, or create a new one? There are several other settings in the Gradle configuration that could use upgrading, so maybe this is a worthwhile effort, but I was hesitant to go that far initially without gauging interest from the maintainers.

**Comment by realbigmitchell** [comment]
> Mitchell, thanks for contributing to this project. The follow-up PR adding IMU seems to have a lot of overlap with this PR, so let's merge this one before I review the next.

Thanks for your review Alexander! I responded to your comments and added another commit.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/AgnssFragment.java [hunk 5]
```diff
@@ -39,12 +39,12 @@
...
   private TextView mLogView;
   private ScrollView mScrollView;
-  private SensorFusionContainer mGpsContainer;
+  private MeasurementProvider mGpsContainer;
   private AgnssUiLogger mUiLogger;
 
   private final AgnssUIFragmentComponent mUiComponent = new AgnssUIFragmentComponent();
 
-  public void setGpsContainer(SensorFusionContainer value) {
+  public void setGpsContainer(MeasurementProvider value) {
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/AgnssFragment.java:47]: Please rename this method to `setMeasurementProvider`.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/AgnssFragment.java [hunk 6]
```diff
@@ -39,12 +39,12 @@
   public static final String TAG = ":AgnssFragment";
   private TextView mLogView;
   private ScrollView mScrollView;
-  private SensorFusionContainer mGpsContainer;
+  private MeasurementProvider mGpsContainer;
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/AgnssFragment.java:42]: Please rename this field to `mMeasurementProvider`.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java [hunk 7]
```diff
@@ -53,7 +54,7 @@
   private boolean firstTime = true;
 
   GoogleApiClient mGoogleApiClient;
-  private final List<SensorFusionListener> mLoggers;
+  private final List<MeasurementListener> mLoggers;
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java:57]: I think it'd be more appropriate to rename this to `mListeners`.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java [hunk 8]
```diff
@@ -62,7 +63,7 @@
         @Override
         public void onProviderEnabled(String provider) {
           if (mLogLocations) {
-            for (SensorFusionListener logger : mLoggers) {
+            for (MeasurementListener logger : mLoggers) {
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java:66]: Likewise this can be `listener : mListeners`.

**Review by alexanderkuznetsov96 — CHANGES_REQUESTED** [review]
A couple more nits, otherwise LGTM!

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java (hunk 1 above)
Comment (alexanderkuznetsov96) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionListener.java]: Oh okay, feel free to leave as is.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java (hunk 2 above)
Comment (alexanderkuznetsov96) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SensorFusionContainer.java]: Please feel free to make that upgrade in the future, but you're right, we can keep that outside the scope of this PR. If you'd like to push that change in the future, I'd be happy to review it.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java (hunk 8 above)
Comment (realbigmitchell) [reply GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java]: Done!

**Comment by realbigmitchell** [comment]
> A couple more nits, otherwise LGTM!

Done! New commit added.

**Review by alexanderkuznetsov96 — APPROVED** [review]

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MainActivity.java [hunk 9]
```diff
@@ -152,7 +153,7 @@ private synchronized void buildGoogleApiClient() {
             .enableAutoManage(this, this)
             .addConnectionCallbacks(this)
             .addOnConnectionFailedListener(this)
-            .addApi(ActivityRecognition.API)
+            .addApi(ActivityRecognition.API).addApi(LocationServices.API)
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MainActivity.java:156]: Do you mind moving the second addApi call to the next line?

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java [hunk 10]
```diff
@@ -105,19 +109,45 @@ public void onLocationChanged(Location location) {
         @Override
         public void onStatusChanged(String provider, int status, Bundle extras) {
           if (mLogLocations) {
-            for (GnssListener logger : mLoggers) {
+            for (MeasurementListener logger : mListeners) {
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/MeasurementProvider.java:112]: For consistency, can you change `logger` to `listener`?

**Review by alexanderkuznetsov96 — APPROVED** [review]
Mitchell, thanks a lot for contributing to this project. I've approved your PR but I left a couple nits, do you mind clearing those up in a follow up PR?

**Comment by mohammedkhider** [comment]
Thanks Mitchel for your appreciated contribution to this open sourced project.
